pkg/k8s/watchers/endpoints.go

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"context"
+	"net/netip"
+	"sync/atomic"
+
+	ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/watchers/resources"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+func (k *K8sWatcher) newEndpointsInit(swgEps *lock.StoppableWaitGroup) {
+	// Use the K8sAPIGroupEndpointV1Core api group to perform the cache sync wait
+	// regardless of what Resource[*Endpoints] is actually using underneath. This is a
+	// temporary workaround until K8sWatcher and "WaitForCacheSync" have been refactored out.
+	apiGroup := resources.K8sAPIGroupEndpointV1Core
+
+	var synced atomic.Bool
+	synced.Store(false)
+
+	k.blockWaitGroupToSyncResources(
+		k.stop,
+		swgEps,
+		func() bool { return synced.Load() },
+		apiGroup,
+	)
+	k.k8sAPIGroups.AddAPI(apiGroup)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	events := k.sharedResources.Endpoints.Events(ctx)
+	go func() {
+		for {
+			select {
+			case <-k.stop:
+				cancel()
+			case event, ok := <-events:
+				if !ok {
+					return
+				}
+				switch event.Kind {
+				case resource.Sync:
+					synced.Store(true)
+				case resource.Upsert:
+					k.K8sEventReceived(apiGroup, resources.MetricEndpointSlice, resources.MetricUpdate, true, false)
```

Comment:
Should we use the underlying resource to manage metrics instead of blindly use EndpointSlice metrics?

Reply:
Changed this up now to figure out which metric name to use in `pkg/k8s/watchers/endpoints.go`. Next steps to further clean this up is to implement metrics in `Resource[T]` itself. This is mostly blocked on the metrics PR. After that we can remove the metrics from `K8sWatchers` that use `Resource[T]` and later remove `K8sWatchers` completely once we've migrated code to use the `Resource[T]` directly.